### PR TITLE
CDPT-2174 RPI: Update identification pages

### PR DESCRIPTION
**Pull request by MeenaModhvadia**
Following a review by content designer the subject photo id and subject address id have been seperated and have their own page.

### CODE REVIEW

File: app/controllers/requests_controller.rb [hunk 1]
```diff
@@ -41,7 +42,7 @@ def new
 
   def show
     redirect_to root_path and return if session[:history].nil?
-
+    # debugger
```
Comment (vertism) [inline app/controllers/requests_controller.rb:45]: Needs removing

File: app/views/requests/_subject_id_check.html.erb [hunk 2]
```diff
@@ -1,8 +1,18 @@
-<%= govuk_panel(html_attributes: { style: "background: #1d70b8; text-align: left; line-height: 1.5em; padding: 25px;" }) do %>
-   <%= @information_request.subject_photo.to_s %>
-   <br>
-   <%= @information_request.subject_proof_of_address.to_s %>
-<% end %>
+ <table class="govuk-table">
```
Comment (vertism) [inline app/views/requests/_subject_id_check.html.erb:1]: The indentation is off

Comment (vertism) [reply app/views/requests/_subject_id_check.html.erb:1]: Also, it would be preferable to use the component - https://govuk-components.netlify.app/components/table/

File: app/views/requests/edit.html.erb [hunk 3]
```diff
@@ -8,6 +8,19 @@
       <h1 class="govuk-heading-xl"><%= content_for?(:title) ? yield(:title) : t("request_form.#{@form.name}") %></h1>
       <%= yield :form_content %>
       <%= f.govuk_submit %>
-    <% end %>
+      <% if content_for?(:text) || t("request_form.#{@form.name}.self_details", default: '').present? %>
+        <details class="govuk-details">
```
Comment (vertism) [inline app/views/requests/edit.html.erb:12]: Please use the component - https://govuk-components.netlify.app/components/details/#output-html-details-with-a-block

File: config/locales/en.yml [hunk 4]
```diff
@@ -233,18 +238,25 @@ en:
...
     subject_id:
-      other: Upload their ID
-      self: Upload your ID
-    subject_id_check: Check your upload
+      other: Upload their photo ID
+      self: Upload your photo ID
+      self_title: If you don't have photo ID
+      self_details: If you are on probation and do not have photo ID, we will accept a letter signed by a probation officer on probation service headed paper. Upload this file here.
+    subject_address:
+      other: Upload your proof of address
```
Comment (vertism) [inline config/locales/en.yml:255]:
```suggestion
      other: Upload their proof of address
```

File: spec/factories/information_requests.rb [hunk 5]
```diff
@@ -39,6 +39,11 @@
       subject_proof_of_address { Rack::Test::UploadedFile.new("spec/fixtures/files/file.jpg", "image/jpg") }
     end
 
+    trait :with_invalid_file_type do
```
Comment (vertism) [inline spec/factories/information_requests.rb:42]: Is this used?

File: spec/models/request_form/subject_id_spec.rb [hunk 6]
```diff
@@ -3,5 +3,32 @@
 RSpec.describe RequestForm::SubjectId, type: :model do
   it_behaves_like "question when requester is not a solicitor"
   it_behaves_like "file upload", :subject_photo
-  it_behaves_like "file upload", :subject_proof_of_address
+
+  describe "#valid_file_type" do
```
Comment (vertism) [inline spec/models/request_form/subject_id_spec.rb:7]: What about if the file type is not valid?

File: spec/models/request_form/subject_id_check_spec.rb [hunk 7]
```diff
@@ -41,11 +42,6 @@
         expect(information_request.subject_photo_id).to be_nil
         expect(information_request.subject_proof_of_address_id).to be_nil
       end
-
```
Comment (vertism) [inline spec/models/request_form/subject_id_check_spec.rb:44]: Why has this been removed?

File: spec/models/request_form/subject_id_check_spec.rb [hunk 8]
```diff
@@ -20,11 +20,12 @@
 
     context "when upload is correct" do
       let(:check) { "yes" }
+      let(:attachment_photo) { create(:attachment, key: "subject_photo") }
 
       it "does not remove the attachment" do
+        expect(information_request.subject_photo).to eq Attachment.find(attachment_photo.id)
+        expect(information_request.subject_proof_of_address).to eq Attachment.find(attachment_address.id)
```
Comment (vertism) [inline spec/models/request_form/subject_id_check_spec.rb:27]: Why are these changes required?

File: spec/requests/subject_spec.rb
[the remark this answers is not in the document]
Comment (MeenaModhvadia) [reply spec/requests/subject_spec.rb:1]: I thought I had in spec/models/request_form/subject_address_spec.rb

File: spec/models/request_form/subject_id_check_spec.rb (hunk 8 above)
Comment (MeenaModhvadia) [reply spec/models/request_form/subject_id_check_spec.rb:27]: undone

File: spec/models/request_form/subject_id_check_spec.rb (hunk 7 above)
Comment (MeenaModhvadia) [reply spec/models/request_form/subject_id_check_spec.rb:44]: undone

File: spec/models/request_form/subject_id_spec.rb (hunk 6 above)
Comment (MeenaModhvadia) [reply spec/models/request_form/subject_id_spec.rb]: added test for invalid file type

File: spec/factories/information_requests.rb (hunk 5 above)
Comment (MeenaModhvadia) [reply spec/factories/information_requests.rb:42]: removed as no longer required

File: spec/models/request_form/subject_id_check_spec.rb (hunk 8 above)
Comment (vertism) [reply spec/models/request_form/subject_id_check_spec.rb:27]: These changes are still here.

File: spec/requests/subject_spec.rb [hunk 9]
```diff
@@ -231,63 +231,157 @@
...
+
+      it_behaves_like "question with back link"
+    end
+  end
+
+  describe "/subject-address" do
+    let(:information_request) { build(:information_request_for_other) }
+    let(:current_step) { "subject-id" }
+    let(:previous_step) { "requester-id" }
+    let(:next_step) { "/subject-address" }
```
Comment (vertism) [inline spec/requests/subject_spec.rb:272]: These are all incorrect

File: spec/requests/subject_spec.rb [hunk 10]
```diff
@@ -231,63 +231,92 @@
...
-        end
+          before do
+            set_session(information_request: information_request.to_hash, current_step: previous_step, history: [previous_step, current_step])
+            get "/#{current_step}"
+          end
 
-        it "saves the associated ID to the session" do
-          patch "/request", params: { request_form: { subject_photo: valid_data, subject_proof_of_address: valid_data } }
-          expect(request.session[:information_request][:subject_photo_id]).to be_an Integer
-          expect(request.session[:information_request][:subject_proof_of_address_id]).to be_an Integer
```
Comment (vertism) [inline spec/requests/subject_spec.rb:270]: I'm not clear why these tests have been changed

File: spec/requests/subject_spec.rb [hunk 11]
```diff
@@ -231,63 +231,92 @@
...
       end
 
       it_behaves_like "question with back link"
     end
+  end
+
+  describe "/subject-address" do
+    let(:information_request) { build(:information_request_for_other) }
+    let(:current_step) { "subject-address" }
+    let(:previous_step) { "requester-id" }
```
Comment (vertism) [inline spec/requests/subject_spec.rb:271]: wrong previous step?

File: spec/requests/subject_spec.rb [hunk 12]
```diff
@@ -231,63 +231,92 @@
...
+        context "when returning to upload your proof of address page" do
+          let(:proof_of_address_upload) { create(:attachment) }
+          let(:information_request) { build(:information_request_by_friend, subject_proof_of_address_id: proof_of_address_upload.id) }
+
+          before do
+            set_session(information_request: information_request.to_hash, current_step: previous_step, history: [previous_step, current_step])
+            get "/#{current_step}"
+          end
+
+          it "doesn't require repeat photo id upload" do
```
Comment (vertism) [inline spec/requests/subject_spec.rb:312]: wrong test name?

File: app/controllers/requests_controller.rb [hunk 13]
```diff
@@ -223,13 +225,15 @@ def next_step
 
   def previous_step
     redirect = nil
-    index = current_index
+    steps_back = @form&.back.present? || @form&.back.nil? ? 1 : 2
```
Comment (vertism) [inline app/controllers/requests_controller.rb:228]: This doesn't seem to make sense.

File: app/views/requests/_subject_id_check.html.erb [hunk 14]
```diff
@@ -1,8 +1,29 @@
...
+      <% row.with_cell do %>
+        <%= link_to "Change", "/subject-id", class: "govuk-link" %>
+      <% end %>
+    <% end %>
+
+    <% body.with_row do |row| %>
+      <% row.with_cell(header: true, text: 'Proof of address') %>
+      <% row.with_cell(text: @information_request.subject_proof_of_address.to_s) %>
+      <% row.with_cell do %>
+        <%= link_to "Change", "/subject-id", class: "govuk-link" %>
```
Comment (vertism) [inline app/views/requests/_subject_id_check.html.erb:21]: Is this going to the right place?

File: app/views/requests/edit.html.erb (hunk 3 above)
Comment (vertism) [reply app/views/requests/edit.html.erb:15]: Is there a reason not to update this?

File: spec/fixtures/files/file.jpg
Comment (vertism) [inline spec/fixtures/files/file.jpg:1]: What's happened to this file. It doesn't appear to be a valid image anymore.

File: spec/models/request_form/subject_id_check_spec.rb [hunk 15]
```diff
@@ -20,6 +20,7 @@
 
     context "when upload is correct" do
       let(:check) { "yes" }
+      let(:attachment_photo) { create(:attachment, key: "subject_photo") }
```
Comment (vertism) [inline spec/models/request_form/subject_id_check_spec.rb:23]: This doesn't seem to be needed.

File: spec/requests/subject_spec.rb [hunk 16]
```diff
@@ -231,113 +231,189 @@
...
-          expect(response.body).to include("For example an electricity or council tax bill in their name")
+          expect(response.body).to include("Upload their photo ID")
+          expect(response.body).to include("For example, a driving licence or passport")
         end
       end
 
       context "when requesting data for self" do
-        let(:information_request) { build(:information_request_for_self) }
+        let(:photo_upload) { create(:attachment) }
+        let(:information_request) { build(:information_request_by_friend, subject_photo_id: photo_upload.id) }
```
Comment (vertism) [inline spec/requests/subject_spec.rb:241]: The test title is "data for self" but the fixture used is `:information_request_by_friend`

File: spec/requests/subject_spec.rb [hunk 17]
```diff
@@ -231,113 +231,189 @@
...
+          end
         end
-      end
 
-      context "when submitting form with valid data" do
-        it "goes to next step" do
-          patch "/request", params: { request_form: { subject_photo: valid_data, subject_proof_of_address: valid_data } }
-          expect(response).to redirect_to(next_step)
-        end
+        context "when returning to upload your proof of address page" do
```
Comment (vertism) [inline spec/requests/subject_spec.rb:258]: I can see this context has been renamed from "when returning to page", but it now refers to the proof of address which is wrong. It also has a bunch of tests included that were not there before. 

The purpose of this context was to test that once a file has been uploaded, the form can be resubmitted without having to upload another file.

File: app/views/requests/edit.html.erb [hunk 18]
```diff
@@ -5,9 +5,14 @@
       <% content_for(:form_content) { render partial: @form.name, locals: { f: } } %>
       <%= f.govuk_error_summary %>
       <span class="govuk-caption-l"><%= yield :caption %></span>
-      <h1 class="govuk-heading-xl"><%= content_for?(:title) ? yield(:title) : t("request_form.#{@form.name}") %></h1>
+      <h1 class="govuk-heading-xl">
+        <%= content_for?(:title) ? yield(:title) : t("request_form.#{@form.name}") %>
+      </h1>
       <%= yield :form_content %>
       <%= f.govuk_submit %>
+      <% if content_for?(:text) || t("request_form.#{@form.name}.self_details", default: '').present? %>
```
Comment (vertism) [inline app/views/requests/edit.html.erb:13]: Is `content_for?(:text)` necessary?

File: spec/models/information_request_summary_spec.rb [hunk 19]
```diff
@@ -142,19 +142,19 @@
...
+    #   it "returns subject photo id details" do
+    #     expect(summary.subject_id[0]).to include({ key: { text: "Photo ID" } })
+    #     expect(summary.subject_id[0]).to include({ value: { text: "file.jpg, 12.2 KB" } })
+    #   end
+    #
+    #   it "returns subject proof of address details" do
+    #     expect(summary.subject_id[1]).to include({ key: { text: "Proof of address" } })
+    #     expect(summary.subject_id[1]).to include({ value: { text: "file.jpg, 12.2 KB" } })
+    #   end
+    # end
```
Comment (vertism) [inline spec/models/information_request_summary_spec.rb:157]: Why are these commented out?

**Review by vertism — APPROVED** [review]
LGTM
